fix: updates for machines and charms

jameinel opened the pull request:
This is a rollup of a few of the changes that I made towards progressing a 'charms/' endpoint and a 'machines/' endpoint. Some of these were also Heather's changes that we merged into my branch.

We also noticed that all of the ModelGeneration functionality wasn't going to be in 4.0 so we don't need to worry about them.

jameinel committed:
fix: cleanup lint

Even though the Charm structure isn't cohesive yet, don't allow it to
break linting, so that we can work on other things without seeing the
warnings about Charm all the time.

jameinel committed:
fix: add .../machines

This just adds the overall listing of machines, and currently returns
the content that would be listed by calling FullStatus.

hmlanigan committed:
feat: add remove show machine

Add POST for model machines to add a machine. Add GET and DELETE for
model machine-id, to get machine details and delete a machine from a
model.

jameinel committed:
fix: machine creation required fields

Some of the machine attributes said that they were required when they
didn't actually exist. The other required definition was at the wrong
level.

jameinel committed:
fix: bring back CharmInfo and GetCharmURLOrigin

Heather removed these when we intended to change how `juju deploy` and
`juju refresh` operated. But for 25.04 we are planning to not change
that functionality.

jameinel committed:
fix: ModelGeneration and UpgradeSeries are N/A

4.0 will be removing most of that functionality, so we don't have a
reason to port them to the Rest API

jameinel committed:
fix: Metrics are also removed in 4.0

hmlanigan commented on openapi.yaml:
```diff
@@ -656,7 +663,8 @@ paths:
           $ref: "#/components/responses/InternalServerError"
-  /models/{namespace}/{model}/charms/{charm}/resources:
+  /models/{namespace}/{model}/charms/{charm-name}/resources:
+    # TODO: jam 2025-03-04 this needs a revision of some sort, to disambiguate revisions of the resources over time
```
The resource name needs to be there also. Resource names are unique to a charm.

This feels awkward. During a deployment with a local resource to upload, the url makes sense. However once the application is deployed, uploading a new resource based on charm doesn't work. Multiple applications can use the same charm, but not guaranteed they want the same charm blob.

Side note: today, juju does not version uploaded resource blobs. We'd have to write out what it looks like and how it integrates with repository resource revisions.

jameinel replied:
So the resource name should be part of the body of the POST, which I think I've managed with the new syntax for the multipart/form-data.

hmlanigan commented on openapi.yaml:
```diff
@@ -667,16 +675,17 @@ paths:
         To deploy a local charm, first upload charms and resources, getting
         references to them, which can then be supplied when creating an
         application.
```
This description still references charms not resources. (lines 671-677)

hmlanigan commented on openapi.yaml:
```diff
@@ -2732,6 +2891,13 @@ components:
+      required: false
+      schema:
+        type: boolean
+        default: true
```
Why do we want `Wait` to be true here? I would have thought false as with KeepInstance.

jameinel replied:
We discussed this in person, but this is "juju remove-machine --no-wait". I was just trying to avoid the "!NoWait" structures where you start having to think in multiply inverted negatives. (Hence the default to Wait: true).
This isn't "wait for the machine to be deleted before returning" it is "wait for the hooks to be completed successfully before destroying the instance".

jameinel replied:
We *could* make the parameter "wait-for-hooks" if we wanted to be a bit more obvious about it.

manadart replied:
I would make it just that.

jameinel committed:
fix: charm-rev and integration-id

Update Integration to be called IntegrationID so that we can more
clearly disambiguate from the "juju remove-relation app-a app-b" from
the "DELETE .../integrations/10"

For uploading resources, we also need to POST to a revision of a charm.

manadart reviewed (approved):
I think this is OK to move forward with.

jameinel committed:
fix: some lint warnings from swagger.io

There were a couple of linter warnings around required referencing
fields that didn't actually exist. I'm not sure why the existing linter
doesn't catch them, but it is good to clean them up.
